sdc/native/utils.hpp

```diff
@@ -28,335 +28,253 @@
 
-    operator const void* () const { return ptr; }
+    parallel_copy(data, temp.get(), len);
+    reorder(temp.get(), index, len, data);
+}
 
-    data_type_arr data        = {};
-    void*         ptr         = &data;
-    uint64_t      actual_size = 0;
-};
+template<typename I>
+void reorder(void* data, I* index, uint64_t len, uint64_t size)
+{
+    std::unique_ptr<quant[]> temp(new quant[len*size]);
 
-using compare_func = bool (*)(const void*, const void*);
-template<uint64_t size> using exact_void_range = void_range<size, exact_void_data<size>>;
-template<uint64_t size> using _void_range = void_range<size, void_data<size>>;
+    parallel_copy(data, temp.get(), len, size);
+    reorder(temp.get(), index, len, size, data);
+}
 
-template<uint64_t size>
-void swap(exact_void_data<size> a, exact_void_data<size> b)
+template<class I, typename Argsort>
+void sort_by_argsort(void* data, uint64_t len, uint64_t size, compare_func cmp, Argsort argsort)
 {
-    std::swap(*a.ptr, *b.ptr);
+    std::unique_ptr<I[]> index(new I[len]);
+
+    argsort(index.get(), data, len, size, cmp);
+    reorder(data, index.get(), len, size);
 }
 
-template<uint64_t size>
-void swap(void_data<size> a, void_data<size> b)
+template<typename T>
+bool nanless(const T& left, const T& right)
 {
-    auto tmp(a);
-    copy(b, a);
-    copy(tmp, b);
+    return std::less<T>()(left, right);
 }
 
-template<class T>
-T* upper_bound(T* first, T* last, T* value, int size, int item_size, void* compare)
-{
-    T* it = nullptr;
-    auto count = size;
+template<>
+bool nanless<float>(const float& left, const float& right);
 
-    auto less = reinterpret_cast<compare_func>(compare);
+template<>
+bool nanless<double>(const double& left, const double& right);
 
-    if (less)
-    {
-        while (count > 0) {
-            it = first;
-            auto step = count / 2;
-            it = advance(it, step, item_size);
-            if (!less(value, it)) {
-                first = advance(it, 1, item_size);
-                count -= step + 1;
-            }
-            else
-                count = step;
-        }
-    }
-    else
+template<typename T>
+struct less
+{
+    bool operator() (const T& left, const T& right) const
     {
-        while (count > 0) {
-            it = first;
-            auto step = count / 2;
-            it = advance(it, step, item_size);
-            if (!(*value < *it)) {
-                first = advance(it, 1, item_size);
-                count -= step + 1;
-            }
-            else
-                count = step;
-        }
+        return nanless<T>(left, right);
     }
+};
 
-    return first;
-}
+namespace tbb_control
+{
+    void init();
 
-template<>
-void* upper_bound(void* first, void* last, void* value, int size, int item_size, void* compare);
+    tbb::task_arena& get_arena();
 
-tbb::task_arena& get_arena();
+    void set_threads_num(uint64_t);
 
-void set_threads_num(uint64_t);
+    void finalize(void*);
```

Comment:
remove `void*` here, wrap it in lambda with proper signature in moduledef  definition